Add details of ‘confirm email address’ and ‘custom retention period’ (#172)
alphagov/notifications-tech-docs: source/documentation/_api_docs.md

1. [Sign in to GOV.UK Notify](https://www.notifications.service.gov.uk/sign-in).
1. Go to the __Templates__ page and select the relevant email template.
1. Select __Edit__.

> 1. Add a placeholder to the email template using double brackets. For example: "Download your file at: ((link_to_file))"

Author: ```suggestion
1. Add a placeholder to the email template using double brackets. For example: "Download your file at: ((link_to_file))"

Your email should also tell recipients how long the file will be available to download.
```

Reviewer: Hmmm... I _think_ here? Because the other section is optional, but we want services to tell recipients how long the file will be available every time.

Author: Yeah, that’s why I put it here – but it could easily work in both.

The only thing is, they won’t know the default retention period unless they read that optional section – because I moved it out of the intro to avoid repetition.

}
```

> #### Ask recipients to confirm their email address before they can download the file

Author: To be more consistent with other sections of the documentation, should we the headings actually be: 

#### Arguments
##### Personalisation

Instead of:

#### Ask recipients to confirm their email address before they can download the file
##### Opt in

❓

Reviewer: No, I think the way it is now reads better. The `personalisation` argument doesn't just control this feature - it's used for a wholly different purpose, so to say "Arguments -> personalisation" is how you use this feature doesn't seem right.

This security feature is currently opt-in only. From 29 March 2023 it will apply to all new files by default, unless you opt out.

##### Opt in
To ask recipients to confirm their email address, set the `verify_email_before_download` flag to `true`.

You will not need to do this after 29 March 2023.

```json
"personalisation":{
  "first_name": "Amala",

>   "application_date": "2018-01-01",

Author: @samuelhwilliams do we need to repeat the full personalisation argument each time?

Or is it ok to just repeat the `link to file` line?

Reviewer: I think it's OK to repeat the whole thing - but if you feel like it's a bit repetitive/irrelevant I think it's also fine to drop them.

>   "link_to_file": {"file": "file as base64 encoded string"}, {"verify_email_before_download": true}

Reviewer: ```suggestion
  "link_to_file": {"file": "file as base64 encoded string", "verify_email_before_download": true}
```

}
```

##### Opt out (not recommended)
If you do not want to use this security feature, set the `verify_email_before_download` flag to `false` by 29 March 2023.

> 

Author: ```suggestion

You should not opt out if you send files that contain personally identifiable information or sensitive information.
```

Should we add this here or in the intro to this section (line 318)?

Reviewer: I think here is good 👍

Author: Thanks @samuelhwilliams – I might even swap this and the preceding line. What do you think?

> If you do not want to use this security feature, set the `verify_email_before_download` flag to `false` by 10 February 2023.

Author: ```suggestion
If you do not want to use this security feature, set the `verify_email_before_download` flag to `false` by 29 March 2023.
```

```json
"personalisation":{
  "first_name": "Amala",
  "application_date": "2018-01-01",
  "link_to_file": {"file": "file as base64 encoded string", "verify_email_before_download": false}
}
```

#### Choose the length of time that a file is available to download

> Set the number of weeks you want the file to be available in the personalisation argument.

Author: @samuelhwilliams is this a better description?

```suggestion
Set the number of weeks you want the file to be available in the `retention_period` flag.
```

Reviewer: ```suggestion
Set the number of weeks you want the file to be available using the `retention_period` key.
```

I think this might make more sense. FYI, when we say `flags` we generally just mean something with a boolean value (True/False, 0/1). In this case `retention_period` has a string (text) value.

> 

Author: @samuelhwilliams Is it worth being explicit and saying 1 week is the smallest acceptable value, or is that implicit?

Reviewer: We should be clear on the range of acceptable values (1 week to 78 weeks).

Author: ```suggestion

You can choose any value between 1 week and 78 weeks.

```

> If you do not do this, the file will be available for the default retention period of 18 months.
> 
> From 29 March 2023 we will reduce the default retention period to 6 months for all new files.

Author: To be consistent we should display these in weeks.

Can a reviewer please sense-check my numbers?

```suggestion
If you do not do this, the file will be available for the default retention period of 72 weeks (18 months).

From 29 March 2023 we will reduce the default retention period to 24 weeks (6 months) for all new files.
```

Reviewer 2: @samuelhwilliams is correct, so 18 months would be 78 weeks.

This is a thing I am worried about when choosing weeks as the only unit of retention for now - it can be quite confusing for longer retentions, easy to make an error and then telling the end user different time in months than the actual retention would be in weeks.

```json
"personalisation":{
  "first_name": "Amala",
  "application_date": "2018-01-01",

>   "link_to_file": {"file": "file as base64 encoded string"}, {"retention_period": 4w}

Reviewer 2: The quotation marks are important to mark it as a string:

```suggestion
  "link_to_file": {"file": "file as base64 encoded string"}, {"retention_period": "4 weeks"}
```

☺️

}
```

#### Ask recipients to confirm their email address before they can download the file

> This security feature is currently opt-in only. From 10 February 2023 it will apply to all new files by default, unless you opt out.

Author: Date TBC

Author: ```suggestion
This security feature is currently opt-in only. From 29 March 2023 it will apply to all new files by default, unless you opt out.
```

##### Opt in
To ask recipients to confirm their email address, set the `verify_email_before_download` flag to `true`.

> You will not need to do this after 10 February 2023.

Author: Date TBC

Author: ```suggestion
You will not need to do this after 29 March 2023.
```

```json
"personalisation":{
  "first_name": "Amala",
  "application_date": "2018-01-01",
  "link_to_file": {"file": "file as base64 encoded string"}, {"verify_email_before_download": true}
}
```

##### Opt out

> If you do not want recipients to confirm their email address, set the `verify_email_before_download` flag to `false` by 10 February 2023.

Author: Date TBC

> If you do not want recipients to confirm their email address, set the `verify_email_before_download` flag to `false` by 10 February 2023.

Reviewer: Perhaps the api docs aren't the place for guidance/recommendations, but just in case - is it worth us highlighting that we don't recommend they do this?

Author: We could do it subtly, for example:

```suggestion
If you do not want to use this security feature, set the `verify_email_before_download` flag to `false` by 10 February 2023.
```

Or we could just flat out say: 

`We do not recommend opting out`

Reviewer: I think I'm inclined to take the hardline approach :) Then it's super clear that they're actively deciding they want to follow a less-secure approach.

Author: I agree.

For me there’s still a question over whether we should be doing more to encourage users to start using this feature _before_ we make it the default. But that’s a separate issue, and one that adds complexity.

```json
"personalisation":{
  "first_name": "Amala",
  "application_date": "2018-01-01",
  "link_to_file": {"file": "file as base64 encoded string"}, {"verify_email_before_download": false}
}
```

#### Choose the length of time that a file is available to download

Set the number of weeks you want the file to be available in the personalisation argument.

If you do not do this, the file will be available for the default retention period of 18 months.

> From 10 February 2023 we will reduce the default retention period to 6 months for all new files.

Author: Date TBC

Author: ```suggestion
From 29 March 2023 we will reduce the default retention period to 6 months for all new files.
```

```json
"personalisation":{
  "first_name": "Amala",
  "application_date": "2018-01-01",

>   "link_to_file": {"file": "file as base64 encoded string"}, {"retention_period": 4w}

Reviewer: ```suggestion
  "link_to_file": {"file": "file as base64 encoded string"}, {"retention_period": "4w"}
```

I haven't totally settled on whether we'll go with '4w' as the shorthand. For now this is fine, but it might change (we might either just use a number eg `4` with weeks implied, or something more human-friendly like "4 weeks"). Undecided...

Author: Depending on when this is decided, there might be a job to update this in all of the clients while I’m away.

Reviewer: That's fine 👍 I'll let it stew in the back of my head over the weekend and hopefully come Tuesday morning I'll commit one way or the other. 🤞

Author: ```suggestion
  "link_to_file": {"file": "file as base64 encoded string", "retention_period": 4 weeks}
```

Reviewer: We are going for the plain English approach of "4 weeks" - but it does need to be in quotes.

}
```

#### Ask recipients to confirm their email address before they can download the file

This feature is currently opt-in only. From 29 March 2023 it will apply to all new files by default, unless you opt out.

> ##### Opt in

Author: ```suggestion
##### Opt in

```

To ask recipients to confirm their email address, set the `verify_email_before_download` flag to `true`.

You will not need to do this after 29 March 2023.

```json
"personalisation":{
  "first_name": "Amala",
  "application_date": "2018-01-01",
  "link_to_file": {"file": "file as base64 encoded string", "verify_email_before_download": true}
}
```

##### Opt out (not recommended)
You should not opt out if you send files that contain personally identifiable information or sensitive information.

If you do not want to use this security feature, set the `verify_email_before_download` flag to `false` by 29 March 2023.

```json
"personalisation":{
  "first_name": "Amala",
  "application_date": "2018-01-01",
  "link_to_file": {"file": "file as base64 encoded string", "verify_email_before_download": false}
}
```

#### Choose the length of time that a file is available to download

Set the number of weeks you want the file to be available using the `retention_period` key.

You can choose any value between 1 week and 78 weeks.

> If you do not do this, the file will be available for the default period of 78 weeks.

Author: ```suggestion
If you do not do this, the file will be available for the default period of 78 weeks (18 months).
```

> If you do not do this, the file will be available for the default period of 78 weeks (18 months).

Reviewer 2: should we also mention here about default change in March 2023?

Author: I left this out so we’re not repeating the exact same info twice (it appears in the intro).

To avoid any repetition it should just be:

> If you do not do this, the file will be available for the default period.

But it seems a little unhelpful to not even say how long that is.

Reviewer 2: Seems fair as long as we remember to update it. I think it may become tricky as we get to February/March 2023 - then we rely on them reading the whole thing to get the correct understanding, whereas many people just skim-read the docs.

Author: We can review that nearer the time – we’ll need to update the docs again for the switchover anyway, so we can look at them when we send reminder comms a month or so before.

Reviewer 2: Sounds good 👍🏼

}
```

#### Ask recipients to confirm their email address before they can download the file

This feature is currently opt-in only. From 29 March 2023 it will apply to all new files by default, unless you opt out.

##### Opt in

To ask recipients to confirm their email address, set the `verify_email_before_download` flag to `true`.

You will not need to do this after 29 March 2023.

```json
"personalisation":{
  "first_name": "Amala",
  "application_date": "2018-01-01",
  "link_to_file": {"file": "file as base64 encoded string", "verify_email_before_download": true}
}
```

> ##### Opt out (not recommended)

Author: ```suggestion
##### Opt out (not recommended)

```

This feature is currently opt-in only. From 29 March 2023 it will apply to all new files by default, unless you opt out.

##### Opt in

To ask recipients to confirm their email address, set the `verify_email_before_download` flag to `true`.

You will not need to do this after 29 March 2023.

```json
"personalisation":{
  "first_name": "Amala",
  "application_date": "2018-01-01",
  "link_to_file": {"file": "file as base64 encoded string", "verify_email_before_download": true}
}
```

##### Opt out (not recommended)

You should not opt out if you send files that contain personally identifiable information or sensitive information.

If you do not want to use this security feature, set the `verify_email_before_download` flag to `false` by 29 March 2023.

```json
"personalisation":{
  "first_name": "Amala",
  "application_date": "2018-01-01",
  "link_to_file": {"file": "file as base64 encoded string", "verify_email_before_download": false}
}
```

#### Choose the length of time that a file is available to download

Set the number of weeks you want the file to be available using the `retention_period` key.

> You can choose any value between 1 week and 78 weeks.

Reviewer 3: I wonder if someone might think this means you could put '10 days' as that is between 1 and 78 weeks. We don't explicitely state that it must be x week(s). Equally, maybe we should look for evidence of people not understanding that instruction before adding more content.

Reviewer: I would hope the line above "set the number of weeks" is enough - along with the error message they'll get from the API.

If you do not do this, the file will be available for the default period of 78 weeks (18 months).

```json
"personalisation":{
  "first_name": "Amala",
  "application_date": "2018-01-01",
  "link_to_file": {"file": "file as base64 encoded string", "retention_period": "4 weeks"}
}
```

> 

Reviewer: Ah good spot - I hadn't noticed that error codes/messages table below.

Hopefully something like `Unsupported retention period '<PERIOD>'. Supported periods are '<1-78> weeks'.`? Roughly in line with the file types error message below.

> 

Reviewer 3: I'l leave you and Karl to figure out the content. 

Also need to think about similar for the `verify_email_before_download` param incase someone puts something non boolean.

> 

Reviewer: @karlchillmaid - for your thoughts if you want to be involved in error message writing that developers will see.

As above for an invalid retention period, maybe: `Unsupported value for retention_period '<PERIOD>'. Supported periods are '<1-78> weeks'.`

And maybe for an invalid value for verify_email_before_download: `Unsupported value for verify_email_before_download: '<VALUE>'. Use a boolean true or false value.`

> 

Reviewer: I'm actually not sure of the value of enumerating these type/value validation errors in the documentation, though... But if it's consistent with what we do at the moment and we want to stick with it, sure.

> 

Author: Suggestions

# For an invalid retention period

error.status_code - [TBC]
error.message - `Unsupported value for retention_period '<PERIOD>'. Supported periods are from 1 to 78 weeks.`
How to fix - Choose a period between 1 and 78 weeks

# For an invalid value for verify_email_before_download

error.status_code - [TBC]
error.message - `Unsupported value for verify_email_before_download: '<VALUE>'. Use a boolean true or false value.`
How to fix - Use either True or False
